src/jdkInstaller.ts

```diff
@@ -0,0 +1,73 @@
+import * as path from "path";
+import * as request from "request";
+import * as fs from "fs";
+import * as mkdirp from "mkdirp";
+import * as pcp from "promisify-child-process";
+
+const homedir = require("os").homedir();
+const bin = path.join(homedir, "bin");
+
+function jabbaUrlSuffix(): string {
+    const runnerOs = process.platform;
+    switch (runnerOs.toLowerCase()) {
+        case "linux":
+            return "linux-amd64";
+        case "darwin":
+            return "darwin-amd64";
+        case "win32":
+            return "windows-amd64.exe";
+        default:
+            throw new Error(
+                `unknown runner OS: ${runnerOs}, expected one of Linux, macOS or Windows.`
+            );
+    }
+}
+
+function isWindows(): boolean {
+    return process.platform === "win32";
+}
+
+function jabbaName(): string {
+    if (isWindows()) return "jabba.exe";
+    else return "jabba";
+}
+
+export function installJava(javaVersion: string, jabbaVersion: string, out?: (msg: string) => void) {
+    const jabbaUrl = `https://github.com/shyiko/jabba/releases/download/${jabbaVersion}/jabba-${jabbaVersion}-${jabbaUrlSuffix()}`;
+    const jabba = path.join(bin, jabbaName());
+    return new Promise<void>((res, rej) => {
+        mkdirp(bin, (err, made) => {
+            if (err) {
+                const msg = `${err} ${made}`;
+                console.log(msg);
+                if (out) out(msg);
+                rej(err);
+            }
+            else res();
+        });
+    }).then(() => curl(jabbaUrl, jabba, true)
+        .then(pth => pcp.exec(`${jabba} ls-remote`))
+        .then(out => outputToString(out.stdout).split("\n")
+            .filter((str, i, arr) => str.includes(javaVersion))[0].trim())
+        .then(async java => {
+            if (out) out(`Installing ${java}`);
+            await pcp.exec(`${jabba} install ${java}`);
```

Comment:
Do we display any progress bar to the user while this download happens? It would be great to stream stdout to the Metals `OutputChannel` and make sure that the output channel is visible during the download.

Reply:
Did you meant something like what I've implemented?